mteb/models/cache_wrapper.py

```diff
@@ -0,0 +1,330 @@
+            raise
+
+    def __contains__(self, text: str) -> bool:
+        return self._hash_text(text) in self.hash_to_index
+
+    def __del__(self):
+        self.close()
+
+    def close(self):
+        if hasattr(self, "vectors") and self.vectors is not None:
+            self.vectors.flush()
+            del self.vectors
+            self.vectors = None
+        logger.info(f"Closed TextVectorMap in directory: {self.directory}")
+
+
+class CachedEmbeddingWrapper:
+    def __init__(self, model: Any, cache_path: Union[str, Path]):
+        self._model = model
+        self.cache_path = Path(cache_path)
+        self.cache_path.mkdir(parents=True, exist_ok=True)
+
+        if hasattr(model, "encode_queries") and hasattr(model, "encode_corpus"):
+            self.encode_method = "split"
+            self.query_cache = TextVectorMap(self.cache_path / "query_cache")
+            self.corpus_cache = TextVectorMap(self.cache_path / "corpus_cache")
+            self.query_cache.load(name="query_cache")
+            self.corpus_cache.load(name="corpus_cache")
+            self._wrap_split_encode_methods()
+        elif hasattr(model, "encode"):
+            self.encode_method = "single"
+            self.cache = TextVectorMap(self.cache_path / "cache")
+            self.cache.load(name="cache")
+            self._wrap_single_encode_method()
+        else:
+            logger.error(
+                "Model must have either 'encode_queries' and 'encode_corpus' methods, or a single 'encode' method."
+            )
+            raise ValueError("Invalid model encoding method")
+
+        logger.info(
+            f"Initialized CachedEmbeddingWrapper with {self.encode_method} encoding method"
+        )
+
+    def _wrap_split_encode_methods(self):
+        original_encode_queries = self._model.encode_queries
+        original_encode_corpus = self._model.encode_corpus
+
+        def wrapped_encode_queries(
+            queries: List[str], batch_size: int = 32, **kwargs
+        ) -> np.ndarray:
+            return self._cached_encode(
+                queries, self.query_cache, original_encode_queries, batch_size, **kwargs
+            )
+
+        def wrapped_encode_corpus(
+            corpus: List[str], batch_size: int = 32, **kwargs
+        ) -> np.ndarray:
+            return self._cached_encode(
+                corpus, self.corpus_cache, original_encode_corpus, batch_size, **kwargs
+            )
+
+        self._model.encode_queries = wrapped_encode_queries
+        self._model.encode_corpus = wrapped_encode_corpus
+
+    def _wrap_single_encode_method(self):
+        original_encode = self._model.encode
+
+        def wrapped_encode(
+            texts: List[str], batch_size: int = 32, **kwargs
+        ) -> np.ndarray:
+            return self._cached_encode(
+                texts, self.cache, original_encode, batch_size, **kwargs
+            )
+
+        self._model.encode = wrapped_encode
+
+    def _cached_encode(
```

Comment:
I would just call this `encode`

Reply:
I think that could get confusing with the models `encode` and this encode which is just a wrapper. Is this style or was there a particular other reason? If just style, I think I'd prefer to leave it.